### PR TITLE
Plumb through github.com/kcp-dev/apimachinery/pkg/logicalcluster.LogicalCluster

### PR DESCRIPTION
**Pull request by sttts**
- adding type safety
- abstract logical cluster retrival from objects (to prepare for deprecation of `metadata.clusterName`)
- prepare for switch to path semantics on the higher layers (i.e. open the door for multiple levels for orgs)

### CODE REVIEW

File: pkg/controller/serviceaccount/tokengetter.go [hunk 1]
```diff
@@ -18,12 +18,14 @@ package serviceaccount
 
 import (
 	"context"
+
```
Comment (marun) [inline pkg/controller/serviceaccount/tokengetter.go]: nit: unrelated change

File: pkg/genericcontrolplane/clientutils/multiclusterconfig.go [hunk 2]
```diff
@@ -35,8 +37,9 @@ import (
 	"k8s.io/apiserver/pkg/server"
 	genericapiserver "k8s.io/apiserver/pkg/server"
 	"k8s.io/client-go/rest"
-	_ "k8s.io/component-base/metrics/prometheus/workqueue" // for workqueue metric registration
+	_ "k8s.io/component-base/metrics/prometheus/workqueue"
 	"k8s.io/klog/v2"
+	// for workqueue metric registration
```
Comment (marun) [inline pkg/genericcontrolplane/clientutils/multiclusterconfig.go]: Does this comment make any sense separated from the import? Also, unrelated formatting change.

File: pkg/registry/core/serviceaccount/storage/token.go [hunk 3]
```diff
@@ -89,6 +90,11 @@ func (r *TokenREST) Create(ctx context.Context, name string, obj runtime.Object,
 		gvk := schema.FromAPIVersionAndKind(ref.APIVersion, ref.Kind)
 		switch {
 		case gvk.Group == "" && gvk.Kind == "Pod":
+			if r.pods == nil {
+				// TODO(sttts): kubelet will try to create bound tokens against the pcluster, but those obviously
+				//              will not work against kcp. ¯\_(ツ)_/¯
+				return nil, errors.NewBadRequest("spec.BoundObjectRef.Kind equal to Pod is invalid")
```
Comment (marun) [inline pkg/registry/core/serviceaccount/storage/token.go:95]: nit: unrelated change

File: pkg/registry/core/rest/genericcontrolplane/storage_core.go [hunk 4]
```diff
@@ -24,6 +24,7 @@ import (
 	"k8s.io/apiserver/pkg/registry/rest"
 	genericapiserver "k8s.io/apiserver/pkg/server"
 	serverstorage "k8s.io/apiserver/pkg/server/storage"
+
```
Comment (marun) [inline pkg/registry/core/rest/genericcontrolplane/storage_core.go]: nit: unrelated formatting change

File: pkg/controller/namespace/namespace_controller.go [hunk 5]
```diff
@@ -34,10 +35,10 @@ import (
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/client-go/util/workqueue"
 	"k8s.io/component-base/metrics/prometheus/ratelimiter"
+	"k8s.io/klog/v2"
```
Comment (marun) [inline pkg/controller/namespace/namespace_controller.go]: nit: These kinds of unrelated formatting changes would ideally be submitted separately. Fine to do cleanup, but we don't have to mix it in with functional changes and unnecessarily expand the scope of review.

File: pkg/controller/serviceaccount/tokens_controller.go [hunk 6]
```diff
@@ -41,6 +43,7 @@ import (
 	"k8s.io/client-go/util/workqueue"
 	"k8s.io/component-base/metrics/prometheus/ratelimiter"
 	"k8s.io/klog/v2"
+
```
Comment (marun) [inline pkg/controller/serviceaccount/tokens_controller.go]: nit: unrelated change

File: pkg/registry/core/serviceaccount/storage/token.go [hunk 7]
```diff
@@ -30,6 +30,7 @@ import (
 	"k8s.io/apiserver/pkg/authentication/authenticator"
 	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
 	"k8s.io/apiserver/pkg/registry/rest"
+
```
Comment (marun) [inline pkg/registry/core/serviceaccount/storage/token.go]: nit: unrelated formatting change

Comment (marun) [reply pkg/registry/core/serviceaccount/storage/token.go]: I won't write any more of these, the few I did was to attempt to underscore the need to remain mindful when composing large changesets such as this. I think if we want to provide meaningful oversight to one another it is important to simplify review as much as possible.

File: pkg/serviceaccount/claims.go [hunk 8]
```diff
@@ -47,12 +49,12 @@ type privateClaims struct {
 }
 
 type kubernetes struct {
-	ClusterName string          `json:"clusterName,omitempty"`
-	Namespace   string          `json:"namespace,omitempty"`
-	Svcacct     ref             `json:"serviceaccount,omitempty"`
-	Pod         *ref            `json:"pod,omitempty"`
-	Secret      *ref            `json:"secret,omitempty"`
-	WarnAfter   jwt.NumericDate `json:"warnafter,omitempty"`
+	ClusterName logicalcluster.LogicalCluster `json:"clusterName,omitempty"`
```
Comment (marun) [inline pkg/serviceaccount/claims.go:51]: (No action required) The persistent length of this everywhere maybe suggests consistent use of a shorter package alias?

File: staging/src/k8s.io/apiserver/pkg/endpoints/request/context_cluster.go [hunk 9]
```diff
@@ -79,13 +79,13 @@ func ValidClusterFrom(ctx context.Context) (*Cluster, error) {
 // ClusterNameFrom returns the cluster name from the value of the cluster
 // key on the ctx.
 // If the cluster name is empty, then return an error.
-func ClusterNameFrom(ctx context.Context) (string, error) {
+func ClusterNameFrom(ctx context.Context) (logicalcluster.LogicalCluster, error) {
```
Comment (marun) [inline staging/src/k8s.io/apiserver/pkg/endpoints/request/context_cluster.go:82]: (No action required) Maybe 'LogicalClusterFrom' is more appropriate now?

File: pkg/controller/serviceaccount/tokens_controller.go [hunk 10]
```diff
@@ -310,7 +312,7 @@ func (e *TokensController) syncSecret() {
 		case sa == nil:
 			// Delete token
 			klog.V(4).Infof("syncSecret(%s|%s/%s), service account does not exist, deleting token", secretInfo.clusterName, secretInfo.namespace, secretInfo.name)
-			if retriable, err := e.deleteToken(secretInfo.clusterName, secretInfo.namespace, secretInfo.name, secretInfo.uid); err != nil {
+			if retriable, err := e.deleteToken(logicalcluster.From(secret), secretInfo.namespace, secretInfo.name, secretInfo.uid); err != nil {
```
Comment (marun) [inline pkg/controller/serviceaccount/tokens_controller.go]: Maybe revert this change, since secretIInfo.clusterName is already what you want?

File: staging/src/k8s.io/apiserver/pkg/server/routes/openapi.go [hunk 11]
```diff
@@ -139,8 +140,8 @@ func (p *openAPIServiceProvider) Register() {
 	}
 
 	err = defaultOpenAPIService.RegisterOpenAPIVersionedService(p.path, &clusterAwarePathHandler{
-		clusterName: "",
-		addHandlerForCluster: func(clusterName string, handler http.Handler) {
+		clusterName: logicalcluster.LogicalCluster{},
```
Comment (marun) [inline staging/src/k8s.io/apiserver/pkg/server/routes/openapi.go:143]: (No action required) Maybe standardize on a preferred empty form (e.g. `New("")` or `LogicalCluster{}` or even `NewEmpty()`)?

File: staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/webhook.go [hunk 12]
```diff
@@ -318,6 +318,9 @@ type subjectAccessReviewV1Client struct {
 func (t *subjectAccessReviewV1Client) Create(ctx context.Context, subjectAccessReview *authorizationv1.SubjectAccessReview, opts metav1.CreateOptions) (result *authorizationv1.SubjectAccessReview, statusCode int, err error) {
 	result = &authorizationv1.SubjectAccessReview{}
 
+	bs, _ := json.Marshal(subjectAccessReview)
```
Comment (marun) [inline staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/webhook.go]: nit: Unrelated changes to this file?

File: staging/src/k8s.io/client-go/discovery/discovery_client.go [hunk 13]
```diff
@@ -418,11 +421,11 @@ func ServerPreferredNamespacedResources(d DiscoveryInterface) ([]*metav1.APIReso
 	}), all), err
 }
 
-func (d *DiscoveryClient) clusterAwarePath(path string) string {
-	if d.cluster == "" {
-		return path
+func (d *DiscoveryClient) clusterAwarePath(p string) string {
```
Comment (marun) [inline staging/src/k8s.io/client-go/discovery/discovery_client.go]: (No action required) Maybe avoid unnecessary var name brevity except for index vars and receivers?

**Review by marun — APPROVED** [review]
LGTM

File: pkg/controller/serviceaccount/tokengetter.go (hunk 1 above)
Comment (sttts) [reply pkg/controller/serviceaccount/tokengetter.go]: moved into its own commit

File: pkg/controller/namespace/namespace_controller.go (hunk 5 above)
Comment (sttts) [reply pkg/controller/namespace/namespace_controller.go]: all removed

File: pkg/genericcontrolplane/clientutils/multiclusterconfig.go (hunk 2 above)
Comment (sttts) [reply pkg/genericcontrolplane/clientutils/multiclusterconfig.go]: fixed

File: pkg/registry/core/serviceaccount/storage/token.go (hunk 3 above)
Comment (sttts) [reply pkg/registry/core/serviceaccount/storage/token.go:95]: moved into its own commit

File: pkg/serviceaccount/claims.go (hunk 8 above)
Comment (sttts) [reply pkg/serviceaccount/claims.go:51]: or just add an empty line

File: pkg/registry/core/serviceaccount/storage/token.go (hunk 3 above)
Comment (marun) [reply pkg/registry/core/serviceaccount/storage/token.go:95]: For future, I suggest these kinds of changes be submitted as separate PRs rather than separate commits in a larger PR with unrelated changes. Ideally a PR has changes that are logically consistent to ensure review can focus on the important details.

File: staging/src/k8s.io/apiserver/pkg/server/routes/openapi.go (hunk 11 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/routes/openapi.go:143]: interestingly you cannot use LogicalCluster{} everywhere

File: staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/webhook.go (hunk 12 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/webhook.go]: debugging code
